src/Spout/Writer/WriterAbstract.php

```diff
@@ -199,64 +193,111 @@ protected function throwIfWriterAlreadyOpened($message)
+     * @internal param \Box\Spout\Writer\Common\Entity\Row $row A Row object with cells and styles
+     *          Example $row = (new Row())->addCell('data1');
+     *
+     * @throws SpoutException If anything else goes wrong while writing data
+     * @throws WriterNotOpenedException If this function is called before opening the writer
+     *
      * @api
-     * @return WriterAbstract
-     * @throws \Box\Spout\Writer\Exception\WriterNotOpenedException If this function is called before opening the writer
-     * @throws \Box\Spout\Common\Exception\IOException If unable to write data
-     * @throws \Box\Spout\Common\Exception\SpoutException If anything else goes wrong while writing data
      */
-    public function addRow(array $dataRow)
+    public function addRow($row)
     {
+        if (!is_array($row) && !$row instanceof Row) {
+            throw new InvalidArgumentException('addRow accepts an array with scalar values or a Row object');
+        }
+
+        if (is_array($row) && !empty($row)) {
+            $row = $this->createRowFromArray($row, null);
+        }
+
         if ($this->isWriterOpened) {
-            // empty $dataRow should not add an empty line
-            if (!empty($dataRow)) {
+            if (!empty($row)) {
                 try {
-                    $this->addRowToWriter($dataRow, $this->rowStyle);
+                    $this->applyDefaultRowStyle($row);
+                    $this->addRowToWriter($row);
                 } catch (SpoutException $e) {
                     // if an exception occurs while writing data,
                     // close the writer and remove all files created so far.
                     $this->closeAndAttemptToCleanupAllFiles();
-
                     // re-throw the exception to alert developers of the error
                     throw $e;
                 }
             }
         } else {
             throw new WriterNotOpenedException('The writer needs to be opened before adding row.');
         }
-
         return $this;
     }
 
     /**
+     * @inheritdoc
+     *
+     * @api
+     */
+    public function withRow(\Closure $callback)
+    {
+        return $this->addRow($callback(new Row()));
+    }
+
+    /**
      * Write given data to the output and apply the given style.
      * @see addRow
      *
-     * @api
-     * @param array $dataRow Array of array containing data to be streamed.
+     * @param array|\Box\Spout\Writer\Common\Entity\Row $row The row to be appended to the stream
      * @param Style $style Style to be applied to the row.
-     * @return WriterAbstract
-     * @throws \Box\Spout\Common\Exception\InvalidArgumentException If the input param is not valid
-     * @throws \Box\Spout\Writer\Exception\WriterNotOpenedException If this function is called before opening the writer
-     * @throws \Box\Spout\Common\Exception\IOException If unable to write data
+     * @return WriterInterface
+     * @internal param array $row Array containing data to be streamed.
```

Comment:
You can drop support for that anytime, as 3.0 will contain other breaking changes.